forte/train_preprocessor.py

```diff
@@ -0,0 +1,355 @@
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+from typing import Optional, Dict, Type, Any, Union, Iterator, List
+from texar.torch.data import DataIterator, Batch
+import torch
+from torch import device
+
+from forte.common.configuration import Config
+from forte.data.converter import Converter
+from forte.data.data_pack import DataPack
+from forte.data.data_pack_dataset import DataPackDataSource, \
+    DataPackDataset
+from forte.data.extractor.base_extractor import BaseExtractor
+from forte.data.ontology.core import Entry
+from forte.data.ontology.core import EntryType
+
+logger = logging.getLogger(__name__)
+
+__all__ = [
+    "TrainPreprocessor"
+]
+
+
+class TrainPreprocessor:
+    r"""
+    `TrainPreprocessor` provides the functionality of doing pre-processing work
+    including building vocabulary, extracting the features, batching and
+    padding (optional). The main functionality is provided by its method
+    :meth:`get_train_batch_iterator` which will return an `iterator` over the
+    batch of preprocessed data. Please refer to the documentation of
+    that method for how the pre-processing is done.
+
+    `TrainPreprocessor` will maintain a Config that stores all the configurable
+    parameters for various components.
+
+    `TrainPreprocessor` will also accept a user request. Internally it will
+    parse this user request and store the parsed result.
+
+    Args:
+        pack_generator (Iterator[DataPack]): A generator of
+            :class:`forte.data.data_pack.DataPack`.
+        request (Dict): A request that specifies how to do train pre-processing.
+            Please refer to :meth:`request` for details.
+        config: A `Dict` or :class:`forte.common.configuration.Config` that
+            configs this preprocessor. See :meth:`default_configs` for
+            the defaults.
+
+
+    .. note::
+        For parameters `request`, user does not necessarily need to provide
+        `converter`. If no `converter` is specified, a default converter of
+        type :class:`forte.data.converter.Converter` will be picked.
+    """
+
+    DATA_INPUT = 0
+    DATA_OUTPUT = 1
+
+    def __init__(self,
+                 pack_generator: Iterator[DataPack],
+                 request: Dict,
+                 config: Optional[Union[Config, Dict]] = None):
+        self._config: Config = \
+            Config(config, default_hparams=self.default_configs())
+        self._validate_config()
+
+        self._pack_generator: Iterator[DataPack] = pack_generator
+        self._cached_packs: List[DataPack] = []
```

Comment:
This cache is a bit worrying. Since in a real training job we cannot really cache everything in memory. In our initial meeting, I mention the need for a memory-disk cacher.

We may also need to turn on and off this cacher.

Reply:
I think we have two places need to have cache:
1. Cache parsed data pack produced by `Reader`
2. Cache `Feature` produced by `Extractor`

The `self._cached_packs` is for the 1st case. And I remembered that our initial meeting mainly discussed about the 2nd case and the memory-disk cacher is for 2nd case.

For 1st case, we do missing a memory-disk cacher. Do you mean that if the memory is full, we should serialize unused data pack into disk and later if we need that pack again (e.g. next epoch), we should deserialize it from disk to memory. I guess this needs more careful design and do you want us to do this in this PR?

For 2nd case, we use `texar`  to help us do this and `texar` internally will call pytorch to do this.

Reply:
Yeah we are discussing case 1.

Indeed we can solve this in another PR. But this doesn't have to be complex. We simply can provide 3 modes: no caching, cache in mem, cache in disk. And let the user make the choice.

Reply:
OK